location/services.py

```diff
@@ -64,6 +66,22 @@ def update_or_create(self, data):
         parent_uuid = data.pop('parent_uuid') if 'parent_uuid' in data else None
         # update_or_create(uuid=location_uuid, ...)
         # doesn't work because of explicit attempt to set null to uuid!
+        loc_type = data.get('type', None)
+        if loc_type in ['R', 'D'] \
+                and not self.user.has_perms(
+                    LocationConfig.gql_mutation_create_region_locations_perms
+                ) \
+                and not self.user.is_superuser:
+            raise PermissionDenied(_("unauthorized to create or update region and district"))
+        if loc_type in ['M', 'V'] \
+                and not self.user.has_perms(
+            LocationConfig.gql_mutation_create_region_locations_perms
+        ) \
+                and not self.user.has_perms(
+            LocationConfig.gql_mutation_create_locations_perms
+        ) \
+                and not self.user.is_superuser:
+            raise PermissionDenied(_("unauthorized to create or update municipalities and villages"))
```

Comment:
Can you move this section to separate function? In addition I think this check can be simplified. 
Please note that type is mandatory in your check no need to use .get()
```suggestion
    def _check_permissions(data):
        loc_type = data['type']
        if self.user.is_superuser or self.user.has_perms(
                    LocationConfig.gql_mutation_create_region_locations_perms
                ):
            return True 
        elif self.user.has_perms(
            LocationConfig.gql_mutation_create_locations_perms
        ) and loc_type in ['M', 'V']:
            return True 
        else: 
            raise PermissionDenied(_("unauthorized to create or update locations on %s level") % loc_type)
```